webui/vendor/Bareos/library/Bareos/BSock/BareosBSock.php

```diff
@@ -940,6 +940,73 @@ public function send_command($cmd, $api=0)
         return $result;
     }
 
+    /**
+     * Calls a command on the Bareos Director and returns its result.
+     *
+     * @param $cmd
+     * @return array: Result received from the Bareos Director.
+     */
+    public function call_fullresult_json($cmd)
+    {
+        // should behave similar to
+        // https://github.com/bareos/bareos/blob/master/python-bareos/bareos/bsock/directorconsolejson.py
+        // def call_fullresult(self, command):
+        //     resultstring = super(DirectorConsoleJson, self).call(command)
+        //     data = None
+        //     if resultstring:
+        //         try:
+        //             data = json.loads(resultstring.decode("utf-8"))
+        //         except ValueError as e:
+        //             # in case result is not valid json,
+        //             # create a JSON-RPC wrapper
+        //             data = {"error": {"code": 2, "message": str(e), "data": resultstring}}
+        //             raise bareos.exceptions.JsonRpcInvalidJsonReceivedException(data)
+        //     return data
+        $resultstring = self::send_command($cmd, 2);
+        $data = \Zend\Json\Json::decode($resultstring, \Zend\Json\Json::TYPE_ARRAY);
+        return $data;
+    }
+
+    /**
+     * Calls a command on the Bareos Director and returns its result.
+     *
+     *  If the JSON-RPC result indicates an error
+     *  (contains the ``error`` element),
+     *  an exception will be raised.
+     *
+     * @param $cmd
+     * @return array: The "result" sub-map of the result received from the Bareos Director.
+     * @throw Exception: if JSON-RPC result contains "error" key.
+     * @throw Exception: if JSON-RPC did not contain "result" nor "error" key.
+     */
+    public function call_json($cmd)
+    {
+        // should behave similar to
+        // https://github.com/bareos/bareos/blob/master/python-bareos/bareos/bsock/directorconsolejson.py
+        // def call(self, command):
+        //     json = self.call_fullresult(command)
+        //     if json == None:
+        //         return
+        //     if "result" in json:
+        //         result = json["result"]
+        //     elif "error" in json:
+        //         raise bareos.exceptions.JsonRpcErrorReceivedException(json)
+        //     else:
+        //         raise bareos.exceptions.JsonRpcInvalidJsonReceivedException(json)
+        //     return result
+        $resultstring = self::send_command($cmd, 2);
+        $json = \Zend\Json\Json::decode($resultstring, \Zend\Json\Json::TYPE_ARRAY);
+        if (empty($json)) {
+            return null;
+        } elseif (array_key_exists("result", $json)) {
+            return $json["result"];
+        } elseif (array_key_exists("error", $json)) {
+            throw new \Exception("JsonRpcErrorReceived: " . $resultstring);
+        } else {
+            throw new \Exception("JsonRpcInvalidJsonReceived: " . $resultstring);
```

Comment:
The exception name is a bit weird.  If we receive invalid json, we are actually in case 1 `empty($json)`, as `$json === NULL`.

I guess this happens when we get unexpected (but valid) json.  As python uses the same nomenclature i would not change it though.

Reply:
In case` \Zend\Json\Json::decode` fails because of an invalid json string, it itself raises an `Zend\\Json\\Exception\\RuntimeException` exception. So this is probably fine (except that we do not handle that exception properly, but that is another story).

Reply:
Im a bit confused.  In an earlier comment you said
> When Python fails to convert to JSON, it generates an exception on its own. In PHP this is not the case

So im not quite sure whether it generates an exception or it doesnt.

Im not quite sure how the check the behaviour of ZendJson but from what i can tell it acts like a polyfill for json_decode() which does not throw an exception but instead just returns null.

https://docs.zendframework.com/zend-json/basics/
> By default, the above two calls will proxy to the json_decode() and json_encode() functions of ext/json, which is bundled in default installations of PHP [...]

Reply:
So looking into this it looks like Zend does throw an exception on invalid json. Weird!